### DISCUSSION TITLE
"magick compare" command to resize the reference image to compare with original image

### DISCUSSION THREAD

**Discussion by PearlPig123** [ImageMagick/ImageMagick#6814] [Help] [answered]
### ImageMagick version

7.1.0-61

### Operating system, version and so on

Linux

### Description

I am trying to use magick compare to calculate the similarity score between original image (reference image) and reconstructed image to see if the image quality is degraded by my image processing program. I have a few questions related to how to achieve this properly.

**Context:**
My reconstructed image has different dimension than the original image (for example, the original image is 800x800, the reconstructed image is 400x400).
When I directly compare these two:
```
% magick compare -metric SSIM original.jpeg reconstructed.jpeg NULL: 
0.53204%    
```
The score is very low, which I guess is caused by the different image dimensions.

Then I tried this command and the score is much higher:
```
% magick compare -metric SSIM -resize 800x800 reconstructed.jpeg original.jpeg NULL:
0.837622% 
```
But I am not sure if I am doing this correctly and my questions are:
1. `magick compare` takes two images as input, does the order of them matter for SSIM score calculation? Do we need to put the reference image first and the reconstructed image second?
2. Can `magick compare` take more than 2 images as input so that multiple images can be compared with the reference image in one single commandline?
3. Is this command the correct way to resize the reconstructed image to match the size of the original image? 
```
% magick compare -metric SSIM -resize 800x800 reconstructed.jpeg original.jpeg NULL:
0.837622% 
```

**Comment by fmw42** [comment, marked as answer]
The order does not matter if not doing sub image comparison where you are trying to find the location of a smaller image inside a larger image. When just doing a compare between two equal size images as you have above, the order does not matter.  But if you need to resize one image, then that will introduce some small amount of degradation which would be included in your final score.

magick compare cannot take more than two images.

If you just want to resize one image then put it in parentheses (escaped as needed for your OS). So on Linux,

magick compare -metric SSIM \( -resize 800x800 reconstructed.jpeg \) original.jpeg NULL:

Is the scale of the smaller image, smaller than the larger image?  That is is the smaller image just a crop of the larger image or has it been resized smaller.  If the same scale, then you should do sub image compare to shift the smaller image to best match in the larger image.  See https://imagemagick.org/Usage/compare/#sub-image and https://imagemagick.org/script/compare.php.  The larger image must come first in this type command

**Reply by PearlPig123** [reply]
Hi @fmw42 Thank you for your quick response! To provide more context, I am trying to use the SSIM score to monitor if my image processing program is functioning (not degrading image quality). I have a few follow up questions:
1. Do we always need to make sure the two images compared by `magick compare` is the same dimension for the correctness of the calculation? (From my read, it seems sub-image search doesn't require the two images to be the same dimension)
2. For my image processing program, most of the operations are scaling the input image (like using `magick -resize wxh`), so the reconstructed images are just scaled to a different size with aspect ratio unchanged. There might be format conversion and compression as well depends on the use cases. In this case, is there any other better way to measure the image quality/similarity between reconstructed image and the original image? Is this one good enough `magick compare -metric SSIM ( -resize 800x800 reconstructed.jpeg ) original.jpeg NULL:`?
3. If the two images are in different format and visually different, for example, the original image is a transparent PNG and the reconstructed image is a JPEG with white background (I do alpha blend when I convert PNG to JPEG), how should I adjust the `magick compare` command correspondingly?
4. There are also a small amount of usecases that will crop input images, I guess for those ones I can use sub image search. But there are also some special cases, like: resize an image and pad the border with white pixels, in this case, is there any good way to measure the image quality? I plan to just skip these special cases because I can't figure out a clean way to compare them.

**Reply by fmw42** [reply]
1) If you do not do sub-image search, the two images need to be the same size. Sub-image search finds the best matching location in the larger image matching the same size region as the size of the smaller image. So when using sub-image search the two images need to be different sizes, but have the same scales (size of features).

2) The only other option for equal size images, is -compose difference, to view how different the image are. Then you can get the difference statistics (standard-deviation) to see numerically. But compare is easier to use.

3) I would not suggest trying to compare an opaque and transparent image, unless you can remove the transparency and have similar images. 

4) Crop the padded image and do sub-image search, unless after cropping they are the same size.

**Reply by PearlPig123** [reply]
Thank you for all the details! They are very helpful!
One follow up question on properly remove the transparency using imagemagick:
I currently use OpenCV and implemented [the alpha blend algorithm](https://stackoverflow.com/questions/2049230/convert-rgba-color-to-rgb) to properly transform a transparent PNG to a white background JPEG (because there are cases where a pixel is a transparent blue pixel, if we just directly remove the alpha channel, it becomes a blue pixel instead of a white pixel). I wonder if there is similar flag in imagemagick. I tried commands like 
```
magick transparent.png -background white opaque.jpeg
```
which doesn't work.

**Reply by fmw42** [reply]
If you want white when the alpha is removed where it was transparent, then use

`magick transparent.png -background white -alpha background -alpha off opaque.jpg`

or

`magick transparent.png -background white -flatten opaque.jpg`

**Reply by PearlPig123** [reply]
Actually it is more than that, I want to achieve the effect that the output image is the same visual effect as if we lay the transparent image on a white background (there could be some half transparent RGBA pixels).
I tried `magick transparent.png -background white -flatten opaque.jpg` with a few test images and it seems to be working
:-D

**Comment by snibgo** [comment]
If you want to compare multiple images to one reference image, without re-reading the reference image multiple times, you can do this:
```
magick reference.png -metric SSIM -format "%[distortion]\n" ^
  ( +clone img1.png -compare +write info: +delete ) ^
  ( +clone img2.png -compare +write info: +delete ) ^
  ( +clone img3.png -compare +write info: +delete ) ^
  NULL:
```
This will write a list of numbers to stderr, such as:
```
0.753883
0.777068
0.739994
```
For Windows BAT, double the percent to `%%`. For bash, use `\` instead of `^`.

**Reply by PearlPig123** [reply]
Thank you! This is helpful!
